feat(api): add liquid presence detection to load_instrument()

pmoegenburg opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

This PR adds a boolean argument to the load_instrument ProtocolContext method that enables liquid presence detection for the loaded pipette via an additional pipette state variable.

Closes EXEC-541.

# Test Plan

<!--
Use this section to describe the steps that you took to test your Pull Request.
If you did not perform any testing provide justification why.

OT-3 Developers: You should default to testing on actual physical hardware.
Once again, if you did not perform testing against hardware, justify why.

Note: It can be helpful to write a test plan before doing development

Example Test Plan (HTTP API Change)

- Verified that new optional argument `dance-party` causes the robot to flash its lights, move the pipettes,
then home.
- Verified that when you omit the `dance-party` option the robot homes normally
- Added protocol that uses `dance-party` argument to G-Code Testing Suite
- Ran protocol that did not use `dance-party` argument and everything was successful
- Added unit tests to validate that changes to pydantic model are correct

-->

Added tests to ensure this bool arg can only be populated in API versions 2.20 onward and when using a Flex robot.

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

pmoegenburg committed:
initial implementation

sfoster1 commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -897,6 +898,7 @@ def load_instrument(
                              control <advanced-control>` applications. You cannot
                              replace an instrument in the middle of a protocol being run
                              from the Opentrons App or touchscreen.
+        :param bool liquid_presence_detection: If ``True``, enable liquid presence detection for instrument.
```
This should be noted as only being acceptable if the API version is above some level. It should also default to `None`, so that we can later raise an exception if the API level is below that value and the argument isn't None

sfoster1 commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -931,6 +933,7 @@ def load_instrument(
         instrument_core = self._core.load_instrument(
             instrument_name=checked_instrument_name,
             mount=checked_mount,
+            liquid_presence_detection=liquid_presence_detection,
```
Before passing it in, we should check that if we're below the API level we added it in then the value is `None`. if it is, though, then `None` should become `False` when we pass to lower levels.

sfoster1 commented on api/tests/opentrons/protocol_engine/state/test_pipette_view.py:
```diff
@@ -62,6 +62,7 @@ def get_pipette_view(
     static_config_by_id: Optional[Dict[str, StaticPipetteConfig]] = None,
     flow_rates_by_id: Optional[Dict[str, FlowRates]] = None,
     nozzle_layout_by_id: Optional[Dict[str, Optional[NozzleMap]]] = None,
+    liquid_presence_detection_by_id: Optional[Dict[str, Optional[bool]]] = None,
```
why `Optional[bool]` rather than just `bool`?

pmoegenburg replied:
The LoadPipetteParams liquidPresenceDetection is Optional

sfoster1 reviewed (changes requested):
Couple changes needed to support the API level distinctions (and we should test them also)

pmoegenburg committed:
added api version checking

pmoegenburg committed:
updated syntax

pmoegenburg committed:
refactored

pmoegenburg committed:
added tests

sfoster1 commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -928,9 +930,22 @@ def load_instrument(
+        ):
+            raise ValueError("Liquid presence detection only available on Flex robot.")
+        if (
+            self._api_version < APIVersion(2, 19)
```
oh, this will be 2.20 - 2.19 is being added in 7.3.1

sfoster1 commented on api/src/opentrons/protocol_engine/state/pipettes.py:
```diff
@@ -133,6 +133,7 @@ class PipetteState:
     static_config_by_id: Dict[str, StaticPipetteConfig]
     flow_rates_by_id: Dict[str, FlowRates]
     nozzle_configuration_by_id: Dict[str, Optional[NozzleMap]]
+    liquid_presence_detection_by_id: Dict[str, Optional[bool]]
```
let's make this non-optional. The argument to the `LoadPipette` command is optional, but let's reify that into a bool the same way we do in protocol context to make this just a bool

sfoster1 commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -928,9 +930,22 @@ def load_instrument(
+            self._core.robot_type != "OT-3 Standard"
+            and liquid_presence_detection is not None
+        ):
+            raise ValueError("Liquid presence detection only available on Flex robot.")
```
This should be an enumerated error of some kind, not just a plain value error

sfoster1 commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -928,9 +930,22 @@ def load_instrument(
+            self._api_version < APIVersion(2, 19)
+            and liquid_presence_detection is not None
+        ):
+            raise APIVersionError(
```
This check should happen _before_ the flex-only check, because we want a protocol running on an ot-2 on 2.18 to raise the version error, while an ot-2 on 2.20 would raise the robot error

sfoster1 reviewed (changes requested):
Couple other nits

pmoegenburg committed:
PR review changes

pmoegenburg committed:
test fix

github-actions[bot] committed:
fix(app-testing): snapshot failure capture (#15483)

This PR is an automated snapshot update request. Please review the
changes and merge if they are acceptable or find you bug and fix it.

Co-authored-by: pmoegenburg <pmoegenburg@users.noreply.github.com>

sfoster1 commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -928,9 +931,24 @@ def load_instrument(
+            and liquid_presence_detection is not None
+        ):
+            raise APIVersionError(
+                "Liquid Presence Detection is only supported on Flex protocols in API Version 2.20 and above."
```
```suggestion
                "Liquid Presence Detection is only supported on Flex protocols in API Version 2.20 and above."
```
```suggestion
                "Liquid Presence Detection is only supported in API Version 2.20 and above."
```

sfoster1 commented on api/src/opentrons/protocols/api_support/util.py:
```diff
@@ -37,6 +37,12 @@
 MODULE_LOG = logging.getLogger(__name__)
 
 
+class RobotTypeError(Exception):
```
Looks like we've found some exceptions that need to be `EnumeratedError`s and aren't yet. We should make this exception inherit from [UnsupportedHardwareError](https://github.com/Opentrons/opentrons/blob/217268511518e3074ce4ca661739bc192099dd5e/shared-data/python/opentrons_shared_data/errors/exceptions.py#L968)

sfoster1 commented on api/src/opentrons/protocols/api_support/util.py:
```diff
@@ -37,6 +37,12 @@
+
+    pass
+
+
```
we should probably add a new api version error with a new code in shared-data  for APIVersionError, and make `UnsupportedAPIError` use [ApiRemoved](https://github.com/Opentrons/opentrons/blob/217268511518e3074ce4ca661739bc192099dd5e/shared-data/python/opentrons_shared_data/errors/exceptions.py#L900) or something (but look at how `UnsupportedApiError` is actually used, might be inappropriate). We can do this in a followup though.

sfoster1 reviewed (changes requested):
Couple last small things but looking good

pmoegenburg committed:
PR review updates

pmoegenburg committed:
Merge branch 'EXEC-541-lld-add-load-instrument-option' of https://github.com/Opentrons/opentrons into EXEC-541-lld-add-load-instrument-option

github-actions[bot] committed:
fix(app-testing): snapshot failure capture (#15515)

This PR is an automated snapshot update request. Please review the
changes and merge if they are acceptable or find you bug and fix it.

Co-authored-by: pmoegenburg <pmoegenburg@users.noreply.github.com>

sfoster1 reviewed (dismissed):
Last thing - API verison 2.20 doesn't exist yet. We'll need to add it, which will be pretty easy, but https://github.com/Opentrons/opentrons/pull/15507 should probably get merged first. So we can treat this as done for now and fix it up once that happens in the next day or two.

pmoegenburg committed:
Merge branch 'edge' into EXEC-541-lld-add-load-instrument-option

ryanthecoder reviewed (approved):
lgtm

pmoegenburg committed:
Merge branch 'edge' into EXEC-541-lld-add-load-instrument-option

pmoegenburg committed:
lint fix

pmoegenburg committed:
Merge branch 'edge' into EXEC-541-lld-add-load-instrument-option

pmoegenburg committed:
reverted change after make setup
